core/lsp/PreemptionTaskManager.cc

```diff
@@ -37,7 +37,8 @@ bool PreemptionTaskManager::tryRunScheduledPreemptionTask(core::GlobalState &gs)
     TypecheckEpochManager::assertConsistentThread(
         typecheckingThreadId, "PreemptionTaskManager::tryRunScheduledPreemptionTask", "typechecking thread");
     auto preemptTask = atomic_load(&this->preemptTask);
-    if (preemptTask != nullptr) {
+    if (preemptTask != nullptr &&
+        atomic_compare_exchange_strong(&this->preemptTask, &preemptTask, shared_ptr<Task>(nullptr))) {
```

Comment:
This is a new concurrent bit!

This logic runs when we try to run a preemption task during a slow path. It now uses `atomic_compare_exchange_strong` to safely de-schedule the task without racing with any cancelation that might be happening concurrently.